plugins/tc/tbf/main.go

```diff
@@ -0,0 +1,235 @@
+		if err != nil {
+			return nil, fmt.Errorf("could not serialize prevResult: %v", err)
+		}
+		res, err := version.NewResult(conf.CNIVersion, resultBytes)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse prevResult: %v", err)
+		}
+		conf.RawPrevResult = nil
+		conf.PrevResult, err = current.NewResultFromResult(res)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert result to current version: %v", err)
+		}
+	}
+	// End previous result parsing
+
+	switch {
+	case conf.Burst < 0 || conf.Rate < 0:
+		return nil, fmt.Errorf("rate and burst must be a positive integer")
+	case conf.Burst == 0 && conf.Rate != 0:
+		return nil, fmt.Errorf("if rate is set, burst must also be set")
+	case conf.Rate == 0 && conf.Burst != 0:
+		return nil, fmt.Errorf("if burst is set, rate must also be set")
+	}
+
+	return &conf, nil
+
+}
+
+func getRoutableHostIF(containerIP net.IP) string {
+	routes, err := netlink.RouteGet(containerIP)
+	if err != nil {
+		return ""
+	}
+
+	for _, route := range routes {
+		link, err := netlink.LinkByIndex(route.LinkIndex)
+		if err != nil || link.Attrs().Name == "lo" {
+			continue
+		}
+
+		return link.Attrs().Name
+	}
+
+	return ""
+}
+
+// cmdAdd is called for ADD requests
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	if conf.Rate == 0 && conf.Burst == 0 {
+		return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("must be called as chained plugin")
+	}
+
+	containerIPConfigs := make([]*current.IPConfig, 0, len(conf.PrevResult.IPs))
+
+	for _, ip := range conf.PrevResult.IPs {
+		if ip.Interface == nil {
+			continue
+		}
+		intIdx := *ip.Interface
+		// Every IP is indexed in to the interfaces array, with "-1" standing
+		// for an unknown interface (which we'll assume to be Container-side
+		// Skip all IPs we know belong to an interface with the wrong name.
+		if intIdx >= 0 && intIdx < len(conf.PrevResult.Interfaces) && conf.PrevResult.Interfaces[intIdx].Name != args.IfName {
+			continue
+		}
+		containerIPConfigs = append(containerIPConfigs, ip)
+	}
+
+	if len(containerIPConfigs) == 0 {
+		return fmt.Errorf("got no container IPs")
+	}
+
+	for _, cip := range containerIPConfigs {
+		containerNs, err := ns.GetNS(conf.PrevResult.Interfaces[*cip.Interface].Sandbox)
+		if err != nil {
+			return err
+		}
+		err = containerNs.Do(func(_ ns.NetNS) error {
+			return CreateIngressQdisc(conf.Rate, conf.Burst, conf.PrevResult.Interfaces[*cip.Interface].Name)
+		})
+		if err != nil {
+			return err
+		}
+
+		for pidx, prevIface := range conf.PrevResult.Interfaces {
+			if *cip.Interface == pidx {
+				continue
+			}
+			nonContainerNetNs, err := ns.GetNS(prevIface.Sandbox)
+			if err != nil {
+				return err
+			}
+			err = nonContainerNetNs.Do(func(netNS ns.NetNS) error {
+				nonContainerIfaceName := getRoutableHostIF(cip.Address.IP)
+				if nonContainerIfaceName != "" {
```

Comment:
I *think* that you can assume that this plugin is being called from the "host" namespace.  In other words, the current NS when the plugin starts is the place to put host devices.  That should remove the need for these nested loop and conditionals.

Reply:
You're right - plugins are always called from the host.